internal/statestorage/redis/redispb/player.go

```diff
@@ -0,0 +1,177 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+All of this can probably be done more succinctly with some more interface and
+reflection, this is a hack but works for now.
+*/
+package redispb
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	om_messages "github.com/GoogleCloudPlatform/open-match/internal/pb"
+	"github.com/GoogleCloudPlatform/open-match/internal/statestorage/redis/playerindices"
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gomodule/redigo/redis"
+	log "github.com/sirupsen/logrus"
+)
+
+// Logrus structured logging setup
+var (
+	pLogFields = log.Fields{
+		"app":       "openmatch",
+		"component": "statestorage",
+	}
+	pLog = log.WithFields(pLogFields)
+)
+
+// UnmarshalPlayerFromRedis unmarshals a Player from a redis hash.
+// This can probably be deprecated if we work on getting the above generic enough.
+// The problem is that protobuf message reflection is pretty messy.
+func UnmarshalPlayerFromRedis(ctx context.Context, pool *redis.Pool, player *om_messages.Player) error {
+
+	// Get the Redis connection.
+	redisConn, err := pool.GetContext(context.Background())
+	defer redisConn.Close()
+	if err != nil {
+		pLog.WithFields(log.Fields{
+			"error":     err.Error(),
+			"component": "statestorage",
+		}).Error("failed to connect to redis")
+		return err
+	}
+
+	// Prepare redis command.
+	cmd := "HGETALL"
+	key := player.Id
+	resultLog := pLog.WithFields(log.Fields{
+		"component": "statestorage",
+		"cmd":       cmd,
+		"key":       key,
+	})
+
+	// Run redis command
+	playerMap, err := redis.StringMap(redisConn.Do(cmd, key))
+
+	// Put values from redis into the Player message
+	player.Properties = playerMap["properties"]
+	player.Pool = playerMap["pool"]
+	player.Assignment = playerMap["assignment"]
+	player.Status = playerMap["status"]
+	player.Error = playerMap["error"]
+
+	// TODO: Room for improvement here.
+	if a := playerMap["attributes"]; a != "" {
+		attrsJSON := fmt.Sprintf("{\"attributes\": %v}", a)
+		err = jsonpb.UnmarshalString(attrsJSON, player)
+		if err != nil {
+			resultLog.Error("failure on attributes")
+			resultLog.Error(a)
+		}
+	}
+
+	resultLog.Debug("state storage operation: player unmarshalled")
+	return err
+
+}
+
+// PlayerWatcher makes a channel and returns it immediately.  It also launches an
+// asynchronous goroutine that watches a redis key and returns updates to
+// that key on the channel.
+//
+// The pattern for this function is from 'Go Concurrency Patterns', it is a function
+// that wraps a closure goroutine, and returns a channel.
+// reference: https://talks.golang.org/2012/concurrency.slide#25
+//
+// NOTE: this function will never stop querying Redis during normal operation! You need to
```

Comment:
I suppose there's an underlying behavior in that the available redis pool could potentially bottleneck watchers if concurrency gets too high. Curious what that might look like.

Reply:
Yeah, possible. My reading of the redigo docs leads me to believe that the pool.Get() should block until a connection is available.  If that becomes the case, there are a couple of approaches, but probably the best choice for these high-load 'watch' processes would be a read-only redis slave and a new k8s service name that points at it, and then coding the ability to have the watch processes use that slave while everything else uses the master.